[OMPIRBuilder] Move debug records to correct blocks.

abidh opened the pull request:
Consider the following small OpenMP target region:

```
!$omp target map(tofrom: x)
  x = x + 1
!$omp end target

```
Currently, when compiled with `flang`, it will generate an outlined function like below (with irrelevant bits removed).

```
void @__omp_offloading_10303_14e8afc__QQmain_l13(ptr %0, ptr %1) { entry:
  %2 = alloca ptr, align 8, addrspace(5)
  %3 = addrspacecast ptr addrspace(5) %2 to ptr
...
  br i1 %exec_user_code, label %user_code.entry, label %worker.exit

user_code.entry:
  %5 = load ptr, ptr %3, align 8, !align !19
  br label %omp.region.after_alloca

omp.region.after_alloca:
  br label %outlined.body

outlined.body:
  br label %omp.target

omp.target:
    #dbg_declare(ptr addrspace(5) %2, !20, !DIExpression(), !21)
  ...
  br label %omp.region.cont, !dbg !23

omp.region.cont:
  call void @__kmpc_target_deinit()
  ret void

worker.exit:
  ret void
}
```

Due to how various basic blocks are generated to implement target region for device, the debug record for variable `x` end up in a different block to the location of the variable (`%5` in this case). The backend can drop such debug records.

This PR moves such records in the correct block. If the location field of the debug record is an `Instruction`, it will ensure that debug record is in same block as the instruction. If the location is an `Argument`, the debug record will be moved to the entry block.

Some minor cleanup in the `UpdateDebugRecord` as that brings it to use the similar logic as in `MoveDebugRecordToCorrectBlock`.

abidh committed:
[OMPIRBuilder] Move debug records to correct blocks.

Consider the following small OpenMP target region:

!$omp target map(tofrom: x)
  x = x + 1
!$omp end target

Currently, when compiled with flang, it will generate an outlined
function like below (with irrelevent bits removed).

void @__omp_offloading_10303_14e8afc__QQmain_l13(ptr %0, ptr %1) {
entry:
  %2 = alloca ptr, align 8, addrspace(5)
  %3 = addrspacecast ptr addrspace(5) %2 to ptr
...
  br i1 %exec_user_code, label %user_code.entry, label %worker.exit

user_code.entry:
  %5 = load ptr, ptr %3, align 8, !align !19
  br label %omp.region.after_alloca

omp.region.after_alloca:
  br label %outlined.body

outlined.body:
  br label %omp.target

omp.target:
    #dbg_declare(ptr addrspace(5) %2, !20, !DIExpression(), !21)
  ...
  br label %omp.region.cont, !dbg !23

omp.region.cont:
  call void @__kmpc_target_deinit()
  ret void

worker.exit:                                      ; preds = %entry
  ret void
}

Due to how various basic blocks are generated, the debug record for
variable x end up in a different block to the location of the variable
(%5 in this case). The backend drops such debug records.

This PR moves such records in the correct block. If the location field
of the debug record is an instruction, it will ensure that debug record
is in same block as the instruction. If the location is an argument, the
debug record will be moved to the entry block.

abidh commented:
Ping. I will appreciate any review comments.

abidh commented:
Ping.

timsmith78 commented:
LGTM.  I have applied this commit on top of a recent upstream commit and run a few of our (HPE's) OpenMP tests both with and without target offload.  I do not see any new failures.  I also took one of our offload tests and dumped the LLVM IR with this branch and compared it to the IR generated by the current upstream compiler.  I can clearly see that #dbg_declares are being moved into the same blocks as the allocas, so I believe that this commit is working as intended. 

I do not yet have commit access to approve or officially review this myself, but I would appreciate if someone else who has commit access would push this through as we would like to see the OpenMP debugging experience improve, and this will help.

skatrak reviewed (approved):
I'm no expert on this, but it looks fine to me. If nobody has concerns and it's passing tests I think it can be unblocked from being merged.

abidh committed:
Adjust a few testcases.

The dyn_ptr parameter is now the last one. So adjust the unit test
accordingly. Harden omptarget-debug-var-1.mlir which was expecting that
debug metadata in certain order that made it fragile.

slinder1 commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -8363,29 +8350,54 @@ static void FixupDebugInfoForOutlinedFunction(
-    for (DbgVariableRecord &DVR : filterDbgVars(I.getDbgRecordRange()))
+    for (DbgVariableRecord &DVR : filterDbgVars(I.getDbgRecordRange())) {
       UpdateDebugRecord(&DVR);
+      MoveDebugRecordToCorrectBlock(&DVR);
```
I can see potential maintainability benefits to keeping these steps separate, but it would be more efficient to handle both steps at once, right?

I'd also consider moving the combined step into a `static` function instead of a lambda. It is used only once, and it seems like the lambda is only used to combat rightward-drift, which a `static` function is even better at.

slinder1 replied:
You're right, I'm not sure what performance I was talking about!

The lambdas LGTM

slinder1 reviewed (commented):
LGTM, with some nits

abidh committed:
Handle reivew comments.

abidh committed:
Handle review comments (2).

Poison locations not handled.
